src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -83,14 +105,24 @@ public class GitHubAppCredentials extends BaseStandardCredentials implements Sta
     private String apiUri;
 
     @SuppressFBWarnings(value = "IS2_INCONSISTENT_SYNC", justification = "#withOwner locking only for #byOwner")
+    @Deprecated
     private String owner;
 
+    private RepositoryAccessStrategy repositoryAccessStrategy;
+    private DefaultPermissionsStrategy defaultPermissionsStrategy;
+
+    @NonNull
+    private transient GitHubAppUsageContext context = new GitHubAppUsageContext();
+
     private transient AppInstallationToken cachedToken;
 
     /**
-     * Cache of credentials specialized by {@link #owner}, so that {@link #cachedToken} is preserved.
+     * Caches temporary instances of these credentials created for use in distinct contexts.
+     *
+     * @see #contextualize
+     * @see GitHubAppUsageContext
      */
-    private transient Map<String, GitHubAppCredentials> byOwner;
+    private transient Map<GitHubAppUsageContext, GitHubAppCredentials> cachedCredentials = new ConcurrentHashMap<>();
```

Comment:
We still dynamically create different credentials for each context, but now the contexts are more complicated.